packages/keyring-controller/src/KeyringController.ts

```diff
@@ -2042,6 +2071,34 @@ export class KeyringController extends BaseController<
     this.messagingSystem.publish(`${name}:unlock`);
   }
 
+  /**
+   * Execute the given function after acquiring the controller lock
+   * and save the keyrings to state after it, or rollback to their
+   * previous state in case of error.
+   *
+   * @param fn - The function to execute.
+   * @returns The result of the function.
+   */
+  async #persistOrRollback<T>(fn: MutuallyExclusiveCallback<T>): Promise<T> {
+    return this.#withControllerLock(async ({ releaseLock }) => {
+      const currentSerializedKeyrings = await this.#getSerializedKeyrings();
+
+      try {
+        const callbackResult = await fn({ releaseLock });
+        // State is committed only if the operation is successful
+        await this.#updateVault();
+
+        return callbackResult;
+      } catch (e) {
+        // Keyrings are cleared and restored to their previous state
+        await this.#clearKeyrings({ skipStateUpdate: true });
+        await this.#restoreSerializedKeyrings(currentSerializedKeyrings);
```

Comment:
Hmm. This would destroy then restore each keyring. I am not sure what side-effects this would have for the hardware wallet and snap keyrings. We can check on that.

Reply:
The effect should not be different than locking and unlocking, since we replace keyring instances also in that case

Reply:
We could be more precise checking if the `serialized` version of each keyring matches what we have as backup, to identify keyrings that have been changed and restore only those, but I think the outcome and the concern of substituting instances will be more or less the same

Reply:
I was thinking of performance side-effects, like having to re-establish connections. Or having to re-enter passwords on hardware wallets, that kinda thing. These type of side-effects would be significantly more inconvenient if there was an error triggering it for each call to `persistAllKeyrings` for example, or if the QR keyring repeatedly failed to initialize.

Though, if things are failing that frequently, probably fixing that should be the main concern rather than minimizing the impact of an error. So probably you are right that this is fine.

>We could be more precise checking if the serialized version of each keyring matches what we have as backup, to identify keyrings that have been changed and restore only those

Hmm, interesting idea. If we do decide to take further steps to mitigate the impact of these rollbacks, this sounds like a great way to do it.